upki-fetch/src/main.rs

```diff
@@ -0,0 +1,325 @@
+
+    introduce_manifest(&manifest)?;
+
+    fs::create_dir_all(local).wrap_err("failed to create local directory")?;
+    let plan = form_plan(&manifest, remote_url, local)?;
+
+    log::info!(
+        "Plan formed. {} steps required ({} bytes to download).",
+        plan.steps.len(),
+        plan.download_bytes()
+    );
+
+    for step in plan.steps {
+        step.execute(&client).await?;
+    }
+
+    log::info!("Success.");
+    Ok(())
+}
+
+fn verify(local: &Path) -> Result<(), Report> {
+    let manifest = serde_json::from_reader(
+        File::open(local.join(MANIFEST_JSON))
+            .map(BufReader::new)
+            .wrap_err("cannot open manifest JSON")?,
+    )
+    .wrap_err("cannot parse manifest JSON")?;
+    introduce_manifest(&manifest)?;
+
+    let plan = form_plan(&manifest, "https://.../", local)?;
+
+    match plan.download_bytes() {
+        0 => Ok(()),
+        bytes => Err(eyre!(
+            "fixing the local cache requires downloading {bytes} bytes"
+        )),
+    }
+}
+
+fn introduce_manifest(manifest: &Manifest) -> Result<(), Report> {
+    use chrono::{DateTime, Utc};
+
+    let dt = DateTime::<Utc>::from_timestamp(manifest.generated_at as i64, 0)
+        .map(|dt| dt.to_rfc3339())
+        .ok_or_else(|| Report::msg("manifest has invalid timestamp"))?;
+
+    log::info!("We have manifest generated at {}", dt);
+    log::info!("Comment: {}", manifest.comment);
+    Ok(())
+}
+
+/// Form a plan of how to synchronise with the remote server.
+///
+/// - `manifest` describes the contents of the remote server.
+/// - `remote_url` is the base URL.
+/// - `local` is the path into which files are downloaded.  The caller ensures this exists.
+fn form_plan(manifest: &Manifest, remote_url: &str, local: &Path) -> Result<Plan, Report> {
+    let mut steps = Vec::new();
+
+    // Collect unreferenced files for deletion
+    let mut unreferenced_files = fs::read_dir(local)
+        .wrap_err("failed to read local directory")?
+        .filter_map(|entry| entry.ok())
+        .map(|entry| Path::new(&entry.file_name()).to_owned())
+        .filter(|name| {
+            name.to_string_lossy()
+                .ends_with(".filter")
+                || name
+                    .to_string_lossy()
+                    .ends_with(".delta")
+        })
+        .collect::<HashSet<_>>();
+
+    for filter in &manifest.filters {
+        if !exists_locally(filter, local)? {
+            steps.push(PlanStep::Download {
+                filter: filter.clone(),
+                remote_url: remote_url.to_owned(),
+                local: local.to_owned(),
+            });
+        } else {
+            unreferenced_files.remove(Path::new(&filter.filename));
+        }
+    }
+
+    for filename in unreferenced_files {
+        steps.push(PlanStep::Delete(local.join(filename)));
+    }
+
+    steps.push(PlanStep::SaveManifest {
+        manifest: manifest.clone(),
+        local: local.to_owned(),
+    });
```

Comment:
What's the value in the indirection of making this a step in the plan, rather than doing it outside of the plan?